feat(batch-exports): Add Azure Blob Storage as batch export destination

buildwithmalik opened the pull request:
## Problem

Customers using Azure infrastructure currently can't export PostHog data directly to Azure Blob Storage. This adds Azure Blob as a new batch export destination.

Addresses #41383

This is PR 1 of 2 (backend only). Frontend changes will be in a follow-up PR. Gated behind a feature flag until frontend is ready.

## Changes

Adds `AzureBlob` as a new batch export destination type with a Temporal workflow (`azure-blob-export`) that uses the existing internal S3 staging pipeline, so no new ClickHouse query logic needed.

Credentials are stored via a new `AzureBlobIntegration` class (connection string auth). Added Azurite emulator to docker-compose for local dev.

Test suite covers end-to-end tests for format/compression combinations, error handling, and file splitting.

## How did you test this code?

    # Run Azure Blob tests (requires Azurite)
    docker-compose up -d objectstorage-azure
    pytest products/batch_exports/backend/tests/temporal/destinations/azure_blob/ -v

    # Verify API blocks creation until frontend is ready
    pytest posthog/api/test/batch_exports/test_create.py::test_cannot_create_azure_blob_batch_export_yet -v

Manual testing: Ran full workflow with Temporal workers, created exports via Django shell, verified blobs appear in Azurite with correct format/compression.

### Test Coverage

| Test File | Coverage |
|-----------|----------|
| `test_activity_data_export.py` | Activity-level exports for all model types (events/persons/sessions), format/compression combinations, empty data handling, file splitting. |
| `test_workflow_execution.py` | End-to-end workflow execution for all models, formats, compressions, graceful handling of no data. |
| `test_workflow_error_states.py` | Transient errors, container not found, invalid credentials, cancellation. |
| `test_workflow_with_azure_account.py` | Real Azure account tests (skipped in CI, for manual validation). |
| `test_utils.py` | Unit tests for blob key generation, file extensions, compression extensions, template variable substitution, unsupported format errors |

**Parametrized combinations tested:**
- **Models:** events, persons, sessions
- **Formats:** JSONLines, Parquet  
- **Compression:** None, gzip, brotli, zstd (where supported)
- **Intervals:** hour, day

## Changelog: Is this feature complete?

No - backend only. Frontend changes coming in a fast-follow PR.

buildwithmalik committed:
feat(batch-exports): Add Azure Blob Storage as batch export destination

buildwithmalik commented on products/batch_exports/backend/temporal/destinations/azure_blob_batch_export.py:
```diff
@@ -0,0 +1,448 @@
+            batch_export_id=inputs.batch_export_id,
+            data_interval_start=inputs.data_interval_start,
+            data_interval_end=inputs.data_interval_end,
+            max_record_batch_size_bytes=1024 * 1024 * 60,  # 60MB
```
Matches batch size used by [S3](https://github.com/PostHog/posthog/blob/master/products/batch_exports/backend/temporal/destinations/s3_batch_export.py#L396) and [BigQuery](https://github.com/PostHog/posthog/blob/master/products/batch_exports/backend/temporal/destinations/bigquery_batch_export.py#L963) (60MB).

buildwithmalik commented on products/batch_exports/backend/temporal/destinations/azure_blob_batch_export.py:
```diff
@@ -0,0 +1,448 @@
+            transformer = ParquetStreamTransformer(
+                compression=inputs.compression,
+                include_inserted_at=True,
+                max_file_size_bytes=inputs.max_file_size_mb * 1024 * 1024 if inputs.max_file_size_mb else 0,
```
No default file size limit since Azure blob limits ([190 TiB](https://learn.microsoft.com/en-us/azure/storage/blobs/scalability-targets)) far exceed realistic export sizes. Matches S3 [behavior](https://github.com/PostHog/posthog/blob/master/products/batch_exports/backend/temporal/destinations/s3_batch_export.py#L430).

buildwithmalik commented on products/batch_exports/backend/tests/temporal/conftest.py:
```diff
@@ -519,3 +520,23 @@ async def generate_test_persons_data(ateam, clickhouse_client, data_interval_sta
+    monkeypatch.setattr(
+        "products.batch_exports.backend.temporal.batch_exports.aiokafka.AIOKafkaProducer",
+        lambda **kwargs: mock_producer,
+    )
```
This fixture mocks the Kafka producer for internal telemetry (`try_produce_app_metrics`), not Kafka as a destination. Without this, I saw tests fail trying to connect to kafka:9092 (Docker hostname) from my host machine, even when the actual export succeeds. Kafka destination tests are unaffected since they manage their own connections.

tomasfarias replied:
This has not been an issue for me (nor has anybody in the team mentioned it). I think the issue may be local to your machine. Is `kafka` somewhere in your `/etc/hosts`?

tomasfarias replied:
Ok, this one is not on you, for some reason we have tagged that step as deprecated (see: https://posthog.com/handbook/engineering/manual-dev-setup). 

I'm having a chat with the team about this, but if I had to guess I would say `kafka` not being in your `/etc/hosts` is likely the error and we don't need this fixture.

buildwithmalik commented on docker-compose.base.yml:
```diff
@@ -182,6 +182,16 @@ services:
+        ports:
+            - '10000:10000'
+        volumes:
+            - azurite-data:/data
```
Good callout we don't need persistent storage so i'll remove this to keep it ephemeral similar to MinIO.

buildwithmalik commented on products/batch_exports/backend/temporal/destinations/azure_blob_batch_export.py:
```diff
@@ -0,0 +1,448 @@
+        base_file_name = f"{base_file_name}-{file_number}"
+
+    if compression is not None:
+        file_name = base_file_name + f".{file_extension}.{COMPRESSION_EXTENSIONS[compression]}"
```
In practice, compression values come from the UI/API which validates against `SUPPORTED_COMPRESSIONS` before reaching this code. However, for consistency with the `file_format` validation above and to provide a clearer error message if this is ever called with an invalid value, I'll add similar error handling.

buildwithmalik committed:
fix(batch-exports): Address code review feedback for Azure Blob export

- Remove undefined azurite-data volume from docker-compose configs
- Add try/finally to ensure Azure client cleanup on fixture teardown
- Add non-empty check for expected_session_ids in test assertion
- Add UnsupportedCompressionError with proper error handling

tomasfarias commented on products/batch_exports/backend/temporal/destinations/azure_blob_batch_export.py:
```diff
@@ -0,0 +1,458 @@
+
+
+@workflow.defn(name="azure-blob-export", failure_exception_types=[workflow.NondeterminismError])
+class AzureBlobBatchExportWorkflow(PostHogWorkflow):
```
nit: It is a bit of a convention that the workflow goes at the end of the module, since they all follow a similar structure and they are not very interesting.

tomasfarias commented on products/batch_exports/backend/temporal/destinations/azure_blob_batch_export.py:
```diff
@@ -0,0 +1,448 @@
+            bytes(self.current_buffer),
+            overwrite=True,
+            max_concurrency=self.max_concurrency,
+        )
```
For the future: Does this mean that we cannot call `upload_blob` concurrently and upload parts out of order? This is possible with S3, and it is something we have been testing.

tomasfarias commented on products/batch_exports/backend/tests/temporal/destinations/azure_blob/test_activity_data_export.py:
```diff
@@ -0,0 +1,227 @@
+import uuid
```
nit: The convention is that if there is only one file we just call it `test_activity.py`. `_data_export` is a bit redundant as that's what every batch export does.

buildwithmalik replied:
Done!

tomasfarias commented on products/batch_exports/backend/tests/temporal/destinations/azure_blob/test_activity_data_export.py:
```diff
@@ -0,0 +1,227 @@
+    BatchExportModel(name="events", schema=None),
+    BatchExportModel(name="persons", schema=None),
+    BatchExportModel(name="sessions", schema=None),
+]
```
Please add all the models used in S3 testing here. Also, this usually goes in `utils.py`.

tomasfarias commented on docker-compose.base.yml:
```diff
@@ -182,6 +182,14 @@ services:
+        command: azurite-blob --blobHost 0.0.0.0 --blobPort 10000 --location /data
+        ports:
+            - '10000:10000'
+
```
Generally, we only include services here that are required to run PostHog. This docker-compose file is executed by everyone working at PostHog, multiple times per day/week/year. Moreover, this is only required for batch exports tests, not to run PostHog itself. So I am very much wary of adding anything here.

I don't know what the best solution is. I am leaning towards having a different `docker-compose` (e.g. `docker-compose.batch-exports.yml`) file we can spin up with this service. 

What do you think @rossgray?

tomasfarias commented on products/batch_exports/backend/tests/temporal/destinations/azure_blob/test_workflow_with_azure_account.py:
```diff
@@ -0,0 +1,233 @@
+        async for blob in container_client.list_blobs(name_starts_with=blob_prefix):
+            await container_client.delete_blob(blob.name)
+    finally:
+        await client.close()
```
I'm curious why you have the yield statement inside the try block. Seems to me that deleting blobs is likely what can fail. But the yield itself should not raise.

Also, let's not fail silently, it would be great if you could add a log line here saying we couldn't clean up.

tomasfarias reviewed (changes requested):
I lied, reviewing this now.

Overall, a pretty good work, There are a few things I'd like to discuss before giving this the green light though, see my comments below.

tomasfarias commented:
> Another question, do we want to have this under a feature flag for gradual rollout? Or do we think internal testing (with a real Azure account) should be sufficient? I'm leaning toward internal testing being sufficient since the implementation follows the same patterns as S3 and the blast radius is limited to users who explicitly configure this destination. We can monitor early usage closely after rollout to catch any issues

I think it's better to include a feature flag. I do agree with you that internal testing is sufficient, and that is generally what we do, but a feature flag gives us a cleaner and straight-forward way to enable the destination once testing is done. This is specially relevant given we'll have multiple PRs for this, so a feature flag helps coordinate everything. 

In the most likely scenario, the feature flag will be disabled until we the day flip it on for everyone, and then promptly delete it. This process doesn't take long though. I just purely like the coordination aspect of having a feature flag.

buildwithmalik committed:
refactor(batch-exports): Address code review feedback

- Add feature flag gating for Azure Blob destination
- Extract common utility functions to destinations/utils.py (get_key_prefix, get_manifest_key)
- Remove Azurite from main docker-compose files
- Switch tests to assert_clickhouse_records_in_azure_blob pattern (matching S3/BigQuery/Postgres) to properly account for export pipeline schema transformations

buildwithmalik committed:
fix: Move yield outside try block and add cleanup warning log

buildwithmalik commented:
@tomasfarias 

Changes in latest commits:
- Added a feature flag (azure-blob-batch-exports)
- Extracted get_key_prefix and get_manifest_key to the common `utils.py` file
- Moved Azurite to dedicated docker-compose.batch-exports.yml
- Switched tests to `assert_clickhouse_records_in_azure_blob pattern` that generates expected data using the same Producer the export uses, so both go through identical transformations (matching S3/BigQuery/Postgres).
- Some more minor refactoring

buildwithmalik committed:
refactor(batch-exports): Consolidate S3 and Azure Blob key generation into shared get_object_key function

buildwithmalik committed:
fix(batch-exports): Address code review feedback for Azure Blob tests

buildwithmalik committed:
refactor(batch-exports): Move get_object_key tests to shared test_utils

- Move S3 key generation tests from s3/test_utils.py to shared test_utils.py
- Update README to include batch-exports compose file in docker command

tomasfarias reviewed (approved):
I don't have anything more to add. I'll test this locally, update the migration, and get it merged. Great work here

buildwithmalik committed:
ci(batch-exports): Include docker-compose.batch-exports.yml in CI

Start Azurite container in backend tests for Azure Blob Storage tests.

rossgray reviewed (approved):
Happy with the changes here too. Really good work, especially for a first PR 🎉

tomasfarias committed:
refactor: Move batch exports docker compose to test folder

tomasfarias committed:
chore: Pin azurite image version

tomasfarias committed:
chore: Also update the README about file change

tomasfarias committed:
chore(ci): Update path to batch-exports docker-compose file

tomasfarias committed:
fix: Typing

tomasfarias committed:
fix: Add Azure icon

tomasfarias commented on products/batch_exports/backend/temporal/destinations/azure_blob_batch_export.py:
```diff
@@ -0,0 +1,382 @@
+        except OverBillingLimitError:
+            return
+
+        if inputs.integration_id is None:
```
no, I don't think I will

tomasfarias commented:
Should have fixed all Python typing in latest commit. Frontend CI will fail as Azure is missing an Icon. I'll add that too.

tomasfarias committed:
fix: OpenAPI types

github-actions[bot] committed:
test(storybook): update UI snapshots

tomasfarias commented:
@buildwithmalik aaand shipped! Thanks for your contribution!

Let me know if you wish to take on the last frontend bits to enable this, otherwise we can also pick it up from here. You have already done most of the work.

The plan is to turn the feature flag on for everyone very quickly after the frontend work is finished.
